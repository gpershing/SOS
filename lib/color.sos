struct color = {float r, float g, float b, float a}

alias colors = array color
