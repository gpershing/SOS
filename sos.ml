(* Top-level of the SOS compiler: scan & parse the input,
   check the resulting AST and generate an SAST from it, generate LLVM IR,
   and dump the module 
   Reference: The MicroC compiler *)
(* Written by Sheron *)
 
type action = Ast | Sast | LLVM_IR | Compile

let () =
  let action = ref Compile in
  let set_action a () = action := a in
  let speclist = [
    ("-a", Arg.Unit (set_action Ast), "Print the AST");
    ("-s", Arg.Unit (set_action Sast), "Print the SAST");
    ("-l", Arg.Unit (set_action LLVM_IR), "Print the generated LLVM IR");
    ("-c", Arg.Unit (set_action Compile),
      "Check and print the generated LLVM IR (default)");
  ] in  
  let usage_msg = "usage: ./sos.native [-a|-s|-l|-c] [file.sos]" in
  let channel = ref stdin in
  Arg.parse speclist (fun filename -> channel := open_in filename) usage_msg;
  
  let lexbuf = Lexing.from_channel !channel in
  let ast = Parser.program Scanner.token lexbuf in  
  match !action with
    Ast -> Astprint.basic_print ast
  | _ -> let sast = Semant.check ast in
    match !action with
      Ast     -> ()
    | Sast    -> Sastprint.basic_print sast
    | LLVM_IR -> print_string (Llvm.string_of_llmodule (Codegen.translate sast))
    | Compile -> let m = Codegen.translate sast in
	Llvm_analysis.assert_valid_module m;
	print_string (Llvm.string_of_llmodule m)
