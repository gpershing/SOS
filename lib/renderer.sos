import shape.sos
import color.sos

struct canvas = {width: int, height: int, file_number: int}

startCanvas: (c : canvas) -> void = gl_startRendering(c.width, c.height)

cvoid : () -> void = 0

drawHelper : (point_structs: path, color_structs: array color, numOfPoints: int, i: int,  points: array float, colors: array float) -> void =
    if i >= numOfPoints /*i>=numOfPoints-1?*/
    then cvoid()
    else
        px : float = point_structs[i].x;
        py : float= point_structs[i].y;
        points[2*i] = px;
        points[2*i+1] = py;

        cr : float = color_structs[i].r;
        cg : float = color_structs[i].g;
        cb : float= color_structs[i].b;
        ca : float= color_structs[i].a;
        colors[4*i] = cr;
        colors[4*i+1] = cg;
        colors[4*i+2] = cb;
        colors[4*i+3] = ca;

        drawHelper(point_structs, color_structs, numOfPoints, i + 1, points, colors)

drawPoints : (point_structs : path, color_structs : array color) -> void = 
        numOfPoints : int = point_structs.length;
        points : array float = numOfPoints*2 of [0.0];
        colors : array float = numOfPoints*4 of [0.0];
        
        drawHelper(point_structs, color_structs, numOfPoints, 0,  points, colors);

        gl_drawPoint(points, colors, 2)

drawPath : (point_structs : path, color_structs : array color,  colorMode : int) -> void = 
        numOfPoints : int= point_structs.length;
        points : array float= numOfPoints*2 of [0.0];
        colors : array float= numOfPoints*4 of [0.0];

        drawHelper(point_structs, color_structs, numOfPoints, 0,  points, colors);
        
        gl_drawCurve(points, colors, colorMode)


drawShape : (point_structs : path, color_structs : array color,  colorMode : int, filled : int) ->
    void = 
        numOfPoints : int = point_structs.length;
        points : array float = numOfPoints*2 of [0.0];
        colors : array float = numOfPoints*4 of [0.0];

        drawHelper(point_structs, color_structs, numOfPoints, 0,  points, colors);

        gl_drawShape(points, colors, colorMode, filled)


endCanvas : (c : canvas) -> void = gl_endRendering(c.width, c.height, c.file_number)
