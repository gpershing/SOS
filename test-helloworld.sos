// for import:
import color.sos
c1 : color = {255.0, 0.0, 0.0, 0.8}
c2 : color = {0.0, 255.0, 0.0, 0.8}
color_arr : colors = [c1, c2]
printf(c1.r)
// for using OpenGL call directly, without our stdlibs
p : array float = [0.1, 0.0, 0.5, 0.0]
c : array float = [255.0, 0.0, 0.0, 0.8, 0, 255.0, 0.0, 0.8]
gl_startRendering()
gl_drawCurve(p, c, 1)
gl_endRendering(1)
