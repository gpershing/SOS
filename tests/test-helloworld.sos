import renderer.sos

a: int = 5
print(a)

p1: point = {-0.9, -0.9}
p2: point = {-0.9, -0.7}
p3: point = {-0.7, -0.7}
p4: point = {-0.7, -0.9}
point_arr : path = [p1, p2, p3, p4]

c1 : color = {255.0, 0.0, 0.0, 0.8}
c2 : color = {0.0, 255.0, 0.0, 0.8}
c3 : color = {0.0, 0.0, 255.0, 0.8}
c4 : color = {100.0, 100.0, 0.0, 0.8}
color_arr : colors = [c1, c2, c3, c4]

canvas1 : canvas = {400, 400, 0}

startCanvas(canvas1)
drawShape(point_arr, color_arr, 0, 1)
endCanvas(canvas1)

canvas2 : canvas = {400, 400, 1}

startCanvas(canvas2)
drawShape(point_arr, color_arr, 0, 1)
endCanvas(canvas2)
