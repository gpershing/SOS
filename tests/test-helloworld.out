5
startRendering...
endRendering...
startRendering...
endRendering...
